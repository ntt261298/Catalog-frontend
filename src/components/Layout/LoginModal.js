import React from 'react';
import { connect } from 'react-redux';
import {
  Button, Modal, ModalHeader, ModalBody, ModalFooter,
} from 'reactstrap';
import { PropTypes } from 'prop-types';
import { showSignup, hideModal } from '../../actions/app';
import { onLogin } from '../../actions/user';
import { errMessage, successMessage } from '../../utils/messages';


export class LoginModal extends React.Component {
  state = {
    username: '',
    password: '',
  }

  onChangeInput = (e) => {
    this.setState({
      [e.target.name]: e.target.value,
    });
  }

  moveToSignup = () => {
    this.props.hideModal();
    this.props.showSignup();
  }

  onLogin = async () => {
    const { username, password } = this.state;
    if (!username) {
      errMessage('Username must not be blank');
      return;
    }
    if (password.length < 6) {
      errMessage('Password must have at least 6 characters');
      return;
    }
    this.props.onLogin(this.state.username, this.state.password)
      .then((message) => {
        successMessage(message);
        this.props.hideModal();
      })
      .catch(err => errMessage(err));
  }

  render() {
    const { modal, hideModal } = this.props;
    const isLoginModal = modal === 'login';
    return (
      <div>
        <Modal isOpen={isLoginModal} toggle={hideModal}>
          <ModalHeader toggle={hideModal}>Login</ModalHeader>
          <ModalBody>
            <div>
              <label htmlFor="username">
                Username:
                <input type="text" name="username" onChange={this.onChangeInput} />
              </label>
            </div>
            <div>
              <label htmlFor="password">
                Password:
                <input type="password" name="password" onChange={this.onChangeInput} />
              </label>
            </div>
            <p onClick={this.moveToSignup}>Create new account?</p>
          </ModalBody>
          <ModalFooter>
            <Button color="primary" onClick={this.onLogin}>Login</Button>
          </ModalFooter>
        </Modal>
      </div>
    );
  }
}

LoginModal.propTypes = {
  modal: PropTypes.string.isRequired,
  showSignup: PropTypes.func.isRequired,
  hideModal: PropTypes.func.isRequired,
  onLogin: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  modal: state.app.modal,
});

export default connect(mapStateToProps, {
  showSignup, hideModal, onLogin,
})(LoginModal);
