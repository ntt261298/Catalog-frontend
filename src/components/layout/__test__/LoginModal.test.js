import React from 'react';
import { shallow } from '../../../enzyme';
import { LoginModal } from '../LoginModal';

describe('src/components/layout/LoginModal.js', () => {
  let wrapper;
  let props;
  let loginButton;
  let modalHeader;
  let newAccountParam;
  let userInput;

  const update = () => {
    loginButton = wrapper.find('Button');
    modalHeader = wrapper.find('ModalHeader');
    newAccountParam = wrapper.find('p');
    userInput = wrapper.find('input[name="username"]');
  };

  const setup = () => {
    wrapper = shallow(
      <LoginModal {...props} />,
    );
    update();
  };

  beforeEach(() => {
    props = {
      modal: 'login',
      onLogin: jest.fn(),
      moveToSignup: jest.fn(),
      onChangeInput: jest.fn(),
      hideModal: jest.fn(),
      showSignup: jest.fn(),
      toggle: jest.fn(),
    };
  });

  it('should render login modal', () => {
    setup();
    // Expect the wrapper object to be defined
    expect(wrapper).toMatchSnapshot();
  });

  it('should call onLogin when login button is clicked', () => {
    setup();
    loginButton.simulate('click');
    expect(props.onLogin).toHaveBeenCalled();
  });

  it('should call hideModal when close button is clicked', () => {
    setup();
    modalHeader.simulate('click');
    expect(props.hideModal).toHaveBeenCalled();
  });

  it('should call showSignup when new account param is clicked', () => {
    setup();
    newAccountParam.simulate('click');
    expect(newAccountParam).toHaveLength(1);
    expect(props.moveToSignup).toHaveBeenCalled();
  });

  it('should call onChangeInput when user\'s input is changed', () => {
    setup();
    userInput.simulate('change', {
      target: {
        name: '123456',
      },
    });
    expect(props.onChangeInput).toHaveBeenCalled();
  });
});
