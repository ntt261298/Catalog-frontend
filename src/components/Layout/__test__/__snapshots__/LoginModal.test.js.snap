// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/Layout/LoginModal.js should render login modal 1`] = `
<div>
  <Modal
    autoFocus={true}
    backdrop={true}
    backdropTransition={
      Object {
        "mountOnEnter": true,
        "timeout": 150,
      }
    }
    centered={false}
    fade={true}
    isOpen={true}
    keyboard={true}
    modalTransition={
      Object {
        "timeout": 300,
      }
    }
    onClosed={[Function]}
    onOpened={[Function]}
    role="dialog"
    toggle={[Function]}
    zIndex={1050}
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      tag="h5"
      toggle={[Function]}
      wrapTag="div"
    >
      Login
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <div>
        <label
          htmlFor="username"
        >
          Username:
          <input
            name="username"
            onChange={[Function]}
            type="text"
          />
        </label>
      </div>
      <div>
        <label
          htmlFor="password"
        >
          Password:
          <input
            name="password"
            onChange={[Function]}
            type="password"
          />
        </label>
      </div>
      <p
        onClick={[Function]}
      >
        Create new account?
      </p>
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="primary"
        onClick={[Function]}
        tag="button"
      >
        Login
      </Button>
    </ModalFooter>
  </Modal>
</div>
`;
